pageserver/src/tenant/timeline/walreceiver/connection_manager.rs

```diff
@@ -214,6 +228,52 @@ pub(super) async fn connection_manager_loop_step(
                     }
                 }
             } => debug!("Waking up for the next retry after waiting for {time_until_next_retry:?}"),
+
+            Some(()) = async {
+                // Calculating time needed to wait until sending the next discovery request.
+                // Current implementation is conservative and sends discovery requests only when there are no candidates.
+
+                if any_activity {
+                    // No need to send discovery requests if there is an active connection or candidates.
+                    return None;
+                }
+
+                // Waiting for an active wait_lsn request.
+                while wait_lsn_status.borrow().is_none() {
+                    wait_lsn_status.changed().await.ok();
+                }
+
+                // All preconditions met, preparing to send a discovery request.
+                let now = Utc::now().naive_utc();
+                let next_discovery_ts = last_discovery_ts
+                    .checked_add_signed(discovery_request_interval)
+                    .unwrap_or(now);
+
+                if next_discovery_ts > now {
+                    // Prevent sending discovery requests too frequently.
+                    tokio::time::sleep((next_discovery_ts - now).to_std().unwrap_or_default()).await;
+                }
```

Comment:
1. I think the checked_add_signed failing can never happen.
2. Why do we need it to be a `chrono::Utc`, wouldn't `std::time::Instant` work just as well?
3. Did you consider making `last_discovery_ts` an `Option<std::time::Instant>`, makes for a better modelling IMO.